src/rust.rs

```diff
@@ -0,0 +1,265 @@
+    let rust_src_dist_url = format!("https://github.com/esp-rs/rust-build/releases/download/v{}/{}", version, rust_src_dist_file);
+    let llvm_url = format!("https://github.com/espressif/llvm-project/releases/download/{}/{}", llvm_release, llvm_file);
+    let idf_tool_xtensa_elf_clang = format!("{}/{}-{}", get_tool_path("xtensa-esp32-elf-clang".to_string()), llvm_release, arch);
+
+    RustToolchain {
+        arch: arch.to_string(),
+        llvm_release,
+        llvm_arch,
+        artifact_file_extension,
+        version: "1.55.0-dev".to_string(),
+        rust_dist,
+        rust_dist_temp: "/tmp/rust".to_string(),
+        rust_src_dist,
+        rust_src_dist_temp: "/tmp/rust-src".to_string(),
+        rust_src_dist_file,
+        rust_dist_file,
+        rust_dist_url,
+        rust_src_dist_url,
+        rust_installer: get_rust_installer(version).to_string(),
+        destination_dir: format!("{}/.rustup/toolchains/esp", home_dir().unwrap().display().to_string()),
+        llvm_file,
+        llvm_url,
+        idf_tool_xtensa_elf_clang
+    }
+}
+
+#[cfg(windows)]
+fn set_env_variable(key:&str, value:String) {
+    use winreg::{enums::HKEY_CURRENT_USER, RegKey};
+    let hkcu = RegKey::predef(HKEY_CURRENT_USER);
+    let (env, _) = hkcu.create_subkey("Environment").unwrap(); // create_subkey opens with write permissions
+    env.set_value(key, &value).unwrap();
+}
+
+#[cfg(windows)]
+fn update_env_path(value: &str) {
+    let path_key = "PATH";
+    use winreg::{enums::HKEY_CURRENT_USER, RegKey};
+    let hkcu = RegKey::predef(HKEY_CURRENT_USER);
+    let env = hkcu.open_subkey("Environment").unwrap();
+    let env_path:String = env.get_value(path_key).unwrap();
+    if !env_path.contains(&value) {
+        let updated_env_path = format!("{};{}", env_path, value);
+        set_env_variable(path_key, updated_env_path);
+    }
+}
+
+#[cfg(unix)]
+fn update_env_path(value: &str) {
+}
+
+#[cfg(unix)]
+fn set_env_variable(key:&str, value:&str) {
+
+}
+
+
+fn install_rust_toolchain(toolchain:&RustToolchain) {
+    if Path::new(toolchain.destination_dir.as_str()).exists() {
+        println!("Previous installation of Rust Toolchain exist in: {}", toolchain.destination_dir);
+        println!("Please, remove the directory before new installation.");
+    } else {
+
+        // Some platfroms like Windows are available in single bundle rust + src, because install
+        // script in dist is not available for the plaform. It's sufficient to extract the toolchain
+        if toolchain.rust_installer.is_empty() {
+            prepare_package_strip_prefix(&toolchain.rust_dist_url,
+                                         &toolchain.rust_dist_file,
+                                         toolchain.destination_dir.to_string(),
+                                         "esp");
+        } else {
+            prepare_package_strip_prefix(&toolchain.rust_dist_url,
```

Comment:
This is not really working on Mac. What I get is a single `lib` directory, `bin` is completely missing:
```
ll .rustup/toolchains/esp/            
total 0
drwxr-xr-x  3 imarkov  imarkov    96B Sep 30 09:50 lib
```

Listing `lib` gives:
```
ll .rustup/toolchains/esp/lib/rustlib/
total 2136
-rw-r--r--  1 imarkov  imarkov     9B Sep 30 09:50 components <- Notice that this is a file?!
-rw-r--r--  1 imarkov  imarkov   874K Sep 30 09:51 install.log
-rw-r--r--  1 imarkov  imarkov   126K Sep 30 09:51 manifest-rust-src
-rw-r--r--  1 imarkov  imarkov     2B Sep 30 09:50 rust-installer-version
drwxr-xr-x  3 imarkov  imarkov    96B Sep 30 09:50 src
-rwxr-xr-x  1 imarkov  imarkov    27K Sep 30 09:50 uninstall.sh
```